DSharpPlus/Net/RestClient.cs

```diff
@@ -135,21 +136,24 @@ private async Task ExecuteRequestAsync(BaseRestRequest request, RateLimitBucket
 
                         if (delay < new TimeSpan(-TimeSpan.TicksPerMinute))
                         {
-                            request.Discord?.DebugLogger?.LogMessage(LogLevel.Error, "REST", "Failed to retrieve ratelimits. Giving up and allowing next request for bucket.", DateTime.Now);
+                            request.Discord?.Logger?.LogError(BaseDiscordClient.RestEventId, "Failed to retrieve ratelimits - giving up and allowing next request for bucket");
                             bucket._remaining = 1;
                         }
 
                         if (delay < TimeSpan.Zero)
                             delay = TimeSpan.FromMilliseconds(100);
 
-                        request.Discord?.DebugLogger?.LogMessage(LogLevel.Warning, "REST", $"Pre-emptive ratelimit triggered. Waiting until {resetDate:yyyy-MM-dd HH:mm:ss zzz} ({delay:c}).", DateTime.Now);
-                        request.Discord?.DebugLogger?.LogTaskFault(Task.Delay(delay).ContinueWith(_ => this.ExecuteRequestAsync(request, null, null)), LogLevel.Error, "RESET", "Error while executing request: ");
+                        request.Discord?.Logger?.LogWarning(BaseDiscordClient.RestEventId, "Pre-emptive ratelimit triggered - waiting until {0:yyyy-MM-dd HH:mm:ss zzz} ({1:c}).", resetDate, delay);
+                        Task.Delay(delay)
+                            .ContinueWith(_ => this.ExecuteRequestAsync(request, null, null))
+                            .LogTaskFault(request.Discord?.Logger, LogLevel.Error, BaseDiscordClient.RestEventId, "Error while executing request");
+
                         return;
                     }
-                    request.Discord?.DebugLogger?.LogMessage(LogLevel.Debug, "REST", $"Request for {bucket}. Allowing.", DateTime.Now);
+                    request.Discord?.Logger?.LogDebug(BaseDiscordClient.RestEventId, "Request for {0} is allowed", bucket);
                 }
                 else
-                    request.Discord?.DebugLogger?.LogMessage(LogLevel.Debug, "REST", $"Initial request for {bucket}. Allowing.", DateTime.Now);
```

Comment:
You know what I think happens here? A harmless race condition. If you look at the line: `request.Discord?.Logger?.LogDebug(BaseDiscordClient.RestEventId, "Request for {0} is allowed", bucket);` What stands out to you?

2 parts should: `"Request for {0} is allowed"`, and `bucket`. It's a formattable string. At this point, we've checked whether we can proceed. If you spam the bucket concurrently (like I did in this test), the decrements are going to happen at the same time, but sometime after the decrements happen (or during them), the log formatter kicks in. It calls the `ToString()` method on `bucket`, which at this point has allowed and denied some requests, and turns it into a string.

It part, it's caused by the PR, but ultimately it's a display and concurrency issue, one we cannot solve without making the REST client entirely sequential. You can replicate this using a debugger, some logging breakpoints, and the following snippet of code (in an eval or something - testbot will run this as-is): 
```cs
await Task.WhenAll(Enumerable.Range(1, 11).Select(x => Channel.SendMessageAsync(x.ToString())))
```

The ratelimits are going to work normally, since the checks and decrements are atomic. We could try to mitigate this by calling `.ToString()` on the bucket at the call site, rather than deferring it to the logger's formatter, though I don't think it's going to help much, if at all. Ultimately it's a minor problem, which doesn't affect the library's functionality otherwise.

If you have a fix which does not involve locking the queue until current bucket is formatted, feel free to explain it.

Reply:
Appending `.ToString()` to each of the bucket logs seemed to resolve it for me, so I think going with that is our best bet.